debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/logminer/LogMinerHistoryRecorder.java

```diff
@@ -0,0 +1,296 @@
+     * todo check if set "_optimizer_gather_stats_on_conventional_dml"=FALSE will eliminate occasional ORA-01438
+     */
+    @Override
+    public void insertIntoTempTable(BigDecimal scn, String tableName, String segOwner, int operationCode, Timestamp changeTime,
+                                    String transactionId, int csf, String redoSql) {
+        if (metrics.getRecordMiningHistory() && executor != null) {
+            executor.execute(() -> {
+                try {
+                    psInsert.setLong(1, scn.longValue());
+                    psInsert.setString(2, tableName == null ? "" : tableName);
+                    psInsert.setString(3, segOwner == null ? "" : segOwner);
+                    psInsert.setLong(4, operationCode);
+                    psInsert.setTimestamp(5, changeTime);
+                    psInsert.setString(6, transactionId);
+                    psInsert.setLong(7, csf);
+                    psInsert.setString(8, redoSql);
+
+                    psInsert.executeUpdate();
+
+                    metrics.incrementTempHistoryTableRecordsCounter();
+                    metrics.setRecordHistoryQueueCapacity(recordQueueCapacity.get());
+                }
+                catch (SQLException | RejectedExecutionException e) {
+                    if (e instanceof SQLDataException) {
+                        if (e.getMessage().contains("ORA-01438")) {
+                            // retry one more time
+                            try {
+                                psInsert.executeUpdate();
+                                metrics.incrementTempHistoryTableRecordsCounter();
+                                metrics.setRecordHistoryQueueCapacity(recordQueueCapacity.get());
+                                LOGGER.warn("Cannot insert data due to {}. " +
+                                        "SCN: {}, Table name: {}, Segment Owner: {}, operation code: {}, ChangeTime: {}, transaction id: {}, csf: {}, sql: {}, retrying ...",
+                                        e, scn.longValue(), tableName, segOwner, operationCode, changeTime, transactionId, csf, redoSql);
+                                LOGGER.warn("Retry succeeded, ignore previous warning");
+                                return;
+                            }
+                            catch (SQLException e1) {
+                                LOGGER.error("Retry failed: {}", e1);
+                            }
+                        }
+                    }
+                    LOGGER.error("Cannot insert data due to {}. " +
+                            "SCN: {}, Table name: {}, Segment Owner: {}, operation code: {}, ChangeTime: {}, transaction id: {}, csf: {}, sql: {}",
+                            e, scn.longValue(), tableName, segOwner, operationCode, changeTime, transactionId, csf, redoSql);
+                    // todo remove after figuring out what the problem is
```

Comment:
there is an Oracle bug: Doc ID 2620584.1 which recommends a workaround. Instead re just retry.
This is happening once in a few tens of millions of records. I cannot come up with a better solution.
This is not critical, because we would have REDO_SQL in the log and will continue.

Reply:
No problem.  This is no longer a concern for us with the `HistoryRecorder` API refactor.